vortex-io/src/limit.rs

```diff
@@ -0,0 +1,190 @@
+use std::future::Future;
+use std::pin::Pin;
+use std::task::{ready, Context, Poll};
+
+use futures::Stream;
+use futures_util::stream::FuturesUnordered;
+use pin_project::pin_project;
+use tokio::sync::{Semaphore, TryAcquireError};
+
+/// [`Future`] that carries the amount of memory it will require to hold the completed value.
+#[pin_project]
+struct SizedFut<Fut> {
+    #[pin]
+    inner: Fut,
+    size_in_bytes: usize,
+}
+
+impl<Fut: Future> Future for SizedFut<Fut> {
+    // We tag the size in bytes alongside the original output
+    type Output = (Fut::Output, usize);
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let size_in_bytes = self.size_in_bytes;
+        let inner = ready!(self.project().inner.poll(cx));
+
+        Poll::Ready((inner, size_in_bytes))
+    }
+}
+
+/// A [`Stream`] that can work on several simultaneous requests, capping the amount of memory it
+/// uses at any given point.
+///
+/// It is meant to serve as a buffer between a producer and consumer of IO requests, with built-in
+/// backpressure that prevents the producer from materializing more than a specified maximum
+/// amount of memory.
+///
+/// This crate internally makes use of tokio's [Semaphore], and thus is only available with
+/// the `tokio` feature enabled.
+#[pin_project]
+#[cfg(feature = "tokio")]
```

Comment:
we can make `limit` a module with `tokio` submodule that is feature flagged and contains all this code

Reply:
yea you're right, I already did that I just left these in, will remove 👍